### PR TITLE
Refactor mantel code for pearson and spearman to drastically improve memory locality

### PR DESCRIPTION
**Pull request by sfiligoi**
Refactor mantel code for pearson and spearman to drastically improve memory locality. 
Also cache intermediate results that are invariant across permutations.
Introduce a memory-optimized version of permute in DistanceMatrix, since it is used inside mantel.

Addresses issue #1755

3k * 3K matrix benchmarks on a MacBook Pro with Intel i7-8850H CPU:
mantel pearson: master 320s , this PR 2s
mantel spearman: master 2750s , this PR 7s

mantel pearson of EMP sized matrices (25k * 25k) with PR code now take
250s on MacBook Pro with Intel i7-8850H CPU
63s on Linux server with Intel Xeon Gold 6242 CPU 

Please complete the following checklist:

* [x] I have read the guidelines in [CONTRIBUTING.md](https://github.com/biocore/scikit-bio/blob/master/CONTRIBUTING.md).

* [x] I have documented all public-facing changes in [CHANGELOG.md](https://github.com/biocore/scikit-bio/blob/master/CHANGELOG.md).

* [ ] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/biocore/scikit-bio/blob/master/COPYING.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied. **It is your responsibility to disclose code, documentation, or other content derived from external source(s).** If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [x] **This pull request does not include code, documentation, or other content derived from external source(s).**

**Note:** [REVIEWING.md](https://github.com/biocore/scikit-bio/blob/master/REVIEWING.md) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

### CODE REVIEW

**Comment by sfiligoi** [comment]
This PR uses openmp with CYTHON. But the MacOS test environment uses the system clang, which does not support openmp, thus MacOS tests are failing.
Note that the conda-provided clang is openmp enabled (and that's what is used for packaging)

**Comment by wasade** [comment]
That is blazingly fast...

For clang, would it be possible to modify the action script (which would resolve #1758)?

File: skbio/stats/distance/_mantel.py [hunk 1]
```diff
@@ -297,6 +320,153 @@ def mantel(x, y, method='pearson', permutations=999, alternative='two-sided',
...
+        significance. Must be greater than or equal to zero. If zero,
+        statistical significance calculations will be skipped and
+        permuted_stats will be an empty array.
+
+    Returns
+    -------
+    orig_stat : 1D array_like
+        Correlation coefficient of the test.
+    permuted_stats : 1D array_like
+        Permutted correlation coefficients of the test.
```
Comment (wasade) [inline skbio/stats/distance/_mantel.py:446]:
```suggestion
        Permuted correlation coefficients of the test.
```

File: skbio/stats/distance/tests/test_mantel.py [hunk 2]
```diff
@@ -46,6 +54,180 @@ def setUp(self):
...
+                                 [2, 1, 5, 4, 0, 3],
+                                 [4, 1, 0, 5, 2, 3],
+                                 [1, 2, 5, 4, 0, 3],
+                                 [5, 4, 0, 1, 3, 2],
+                                 [3, 0, 1, 5, 4, 2],
+                                 [5, 0, 2, 3, 1, 4]], dtype=np.int)
+
+        ym_normalized = ym/normym
+
+        permuted_stats = np.empty(14, dtype=x_data.dtype)
```
Comment (wasade) [inline skbio/stats/distance/tests/test_mantel.py:166]:
```suggestion
        permuted_stats = np.empty(len(perm_order), dtype=x_data.dtype)
```

File: skbio/stats/distance/tests/test_mantel.py [hunk 3]
```diff
@@ -46,6 +54,180 @@ def setUp(self):
...
+        mantel_perm_pearsonr_cy(x_data, perm_order, xmean, normxm,
+                                ym_normalized, permuted_stats)
+        for i in range(14):
```
Comment (wasade) [inline skbio/stats/distance/tests/test_mantel.py:169]:
```suggestion
        for i in range(len(perm_order)):
```

File: skbio/stats/distance/tests/test_mantel.py [hunk 4]
```diff
@@ -46,6 +54,180 @@ def setUp(self):
...
+    def test_pearsonr_full(self):
+        """
+        Compare the optimized version of pearson mantel
+        with the naive loop implementation
+        """
+        x = DistanceMatrix.read(get_data_path('dm2.txt'))
+        y = DistanceMatrix.read(get_data_path('dm3.txt'))
+
+        np.random.seed(0)
+        orig_stat_fast, permuted_stats_fast = _mantel_stats_pearson(x, y, 12)
```
Comment (wasade) [inline skbio/stats/distance/tests/test_mantel.py:183]: might be nice to put `12` into a variable

File: skbio/stats/distance/tests/test_mantel.py [hunk 5]
```diff
@@ -46,6 +54,180 @@ def setUp(self):
...
+    def test_spearmanr_full(self):
+        """
+        Compare the optimized version of spearman mantel
+        with the naive loop implementation
+        """
+        x = DistanceMatrix.read(get_data_path('dm2.txt'))
+        y = DistanceMatrix.read(get_data_path('dm3.txt'))
+
+        np.random.seed(0)
+        orig_stat_fast, permuted_stats_fast = _mantel_stats_spearman(x, y, 12)
```
Comment (wasade) [inline skbio/stats/distance/tests/test_mantel.py:211]: same

**Review by wasade — APPROVED** [review]
Just some minor requests -- thank you, this is awesome!!!

File: skbio/stats/distance/_cutils.pyx [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+    cdef Py_ssize_t row,col
+    cdef Py_ssize_t vrow
+    cdef Py_ssize_t idx
+
+    for row in prange(out_n-1, nogil=True):
+        vrow = reorder_vec[row]
+        idx = row*(out_n-1) - ((row-1)*row)/2
+        for col in range(out_n-row-1):
+           out_mat_condensed[idx+col] = in_mat[vrow, reorder_vec[col+row+1]]
```
[the remark this answers is not in the document]
Comment (sfiligoi) [reply skbio/stats/distance/_cutils.pyx:119]: The compiler should be able to optimize that out.
Adding another variable is unlikely to be beneficial.

File: skbio/stats/distance/tests/test_mantel.py (hunk 4 above)
Comment (sfiligoi) [reply skbio/stats/distance/tests/test_mantel.py]: Done

File: skbio/stats/distance/tests/test_mantel.py (hunk 5 above)
Comment (sfiligoi) [reply skbio/stats/distance/tests/test_mantel.py]: Done

**Comment by sfiligoi** [comment]
> That is blazingly fast...
> 
> For clang, would it be possible to modify the action script (which would resolve #1758)?

@wasade 
Done.
Both Linux and MacO Python 3.6 tests are passing now.

**Comment by sfiligoi** [comment]
CC @gregcaporaso

**Comment by sfiligoi** [comment]
In case it was not clear, I do not have write privileges, so someone else has to merge.

**Pull request merged**